temporalio/worker/_nexus.py

```diff
@@ -0,0 +1,471 @@
+                        )
+                elif task.HasField("cancel_task"):
+                    task = task.cancel_task
+                    if _task := self._running_tasks.get(task.task_token):
+                        # TODO(nexus-prerelease): when do we remove the entry from _running_operations?
+                        _task.cancel()
+                    else:
+                        logger.debug(
+                            f"Received cancel_task but no running task exists for "
+                            f"task token: {task.task_token}"
+                        )
+                else:
+                    raise NotImplementedError(f"Invalid Nexus task: {task}")
+
+            except temporalio.bridge.worker.PollShutdownError:
+                exception_task.cancel()
+                return
+
+            except Exception as err:
+                raise RuntimeError("Nexus worker failed") from err
+
+    # Only call this if run() raised an error
+    async def drain_poll_queue(self) -> None:
+        while True:
+            try:
+                # Take all tasks and say we can't handle them
+                task = await self._bridge_worker().poll_nexus_task()
+                completion = temporalio.bridge.proto.nexus.NexusTaskCompletion(
+                    task_token=task.task.task_token
+                )
+                completion.error.failure.message = "Worker shutting down"
+                await self._bridge_worker().complete_nexus_task(completion)
+            except temporalio.bridge.worker.PollShutdownError:
+                return
+
+    # Only call this after run()/drain_poll_queue() have returned. This will not
+    # raise an exception.
+    async def wait_all_completed(self) -> None:
+        await asyncio.gather(*self._running_tasks.values(), return_exceptions=True)
+
+    # TODO(nexus-preview): stack trace pruning. See sdk-typescript NexusHandler.execute
+    # "Any call up to this function and including this one will be trimmed out of stack traces.""
+
+    async def _handle_cancel_operation_task(
+        self, request: temporalio.api.nexus.v1.CancelOperationRequest, task_token: bytes
+    ) -> None:
+        """
+        Handle a cancel operation task.
+
+        Attempt to execute the user cancel_operation method. Handle errors and send the
+        task completion.
+        """
+        ctx = CancelOperationContext(
+            service=request.service,
+            operation=request.operation,
+        )
+        temporal_operation_context.set(
+            _TemporalNexusOperationContext(
+                nexus_operation_context=ctx,
+                client=self._client,
+                task_queue=self._task_queue,
+            )
+        )
+        # TODO(nexus-prerelease): headers
+        try:
+            await self._handler.cancel_operation(ctx, request.operation_token)
+        except Exception as err:
+            logger.exception("Failed to execute Nexus cancel operation method")
```

Comment:
If possible, use the contextual logger that has the Nexus info on it for this and start exception

Reply:
This is already using the contextual logger